pkg/cache/cache_test.go

```diff
@@ -121,6 +122,237 @@ var _ = Describe("Multi-Namespace Informer Cache", func() {
+			} else {
+				transformerHits[s] = 1
+			}
+			return i, nil
+		}
+
+		clien, err = client.New(cfg, client.Options{})
+		Expect(err).NotTo(HaveOccurred())
+		err = ensureNamespace(testNamespaceOne, clien)
+		Expect(err).NotTo(HaveOccurred())
+
+		pod := pod.DeepCopy()
+
+		unstructuredPod := &unstructured.Unstructured{}
+		unstructuredPod.SetGroupVersionKind(pod.GroupVersionKind())
+
+		metadataPod := &metav1.PartialObjectMetadata{}
+		metadataPod.SetGroupVersionKind(pod.GroupVersionKind())
+
+		By("creating the informer cache")
+		informerCache, err = cache.New(cfg, cache.Options{
+			TransformFuncByObject: cache.TransformFuncByObject{
+				unstructuredPod: func(i interface{}) (interface{}, error) {
+					return transform("unstructured", i)
+				},
+				metadataPod: func(i interface{}) (interface{}, error) {
+					return transform("metadata", i)
+				},
+				pod: func(i interface{}) (interface{}, error) {
+					return transform("structured", i)
+				},
+			},
+		})
+		Expect(err).NotTo(HaveOccurred())
+
+		informerCacheCtx, informerCacheCancel = context.WithCancel(context.Background())
+		Expect(cfg).NotTo(BeNil())
+
+		err = clien.Create(informerCacheCtx, pod)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("running the cache")
+		go func(ctx context.Context) {
+			defer GinkgoRecover()
+			Expect(informerCache.Start(ctx)).To(Succeed())
+		}(informerCacheCtx)
+		Expect(informerCache.WaitForCacheSync(informerCacheCtx)).To(BeTrue())
+	})
+
+	AfterEach(func() {
+		unstructuredPod := &unstructured.Unstructured{}
+		unstructuredPod.SetGroupVersionKind(pod.GroupVersionKind())
+		unstructuredPod.SetName("test.pod")
+		unstructuredPod.SetNamespace(pod.Namespace)
+
+		clien.Delete(informerCacheCtx, unstructuredPod)
+		informerCacheCancel()
+	})
+
+	It("should transform unstructured objects", func() {
+		By("making sure transform is applied to initial cache listing")
+		podsList := &unstructured.UnstructuredList{
+			Object: map[string]interface{}{
+				"apiVersion": "v1",
+				"kind":       "PodList",
+			},
+		}
+
+		err := informerCache.List(informerCacheCtx, podsList)
+		Expect(err).ToNot(HaveOccurred())
+
+		// Make sure transformer is applied on create
+		Expect(transformerHits["unstructured"]).To(Equal(1))
+
+		By("making sure transform is applied to updated cache listings")
+		unstructuredPod := &podsList.Items[0]
+		unstructuredPod.SetAnnotations(map[string]string{
+			"test": "hello",
+		})
+
+		err = clien.Update(informerCacheCtx, unstructuredPod)
+		Expect(err).ToNot(HaveOccurred())
+
+		podsList = &unstructured.UnstructuredList{
+			Object: map[string]interface{}{
+				"apiVersion": "v1",
+				"kind":       "PodList",
+			},
+		}
+
+		err = informerCache.List(informerCacheCtx, podsList)
+		Expect(err).ToNot(HaveOccurred())
+
+		//!TODO: I do not know a way to force the underlying listerwatcher
+		// to update. Please advise. Sleep here temporarily since it is required
+		// to reliably pass.
+		time.Sleep(1 * time.Second)
```

Comment:
@FillZpp or someone with knowledge: please advise on the best way to wait/trigger listerwatcher events for these tests?

Reply:
You can't, at least not within the integration tests, it happens asynchronously. The only way to do this I think would be to use a fake listwatcher, but that would be a largish refactoring: https://github.com/kubernetes-sigs/controller-runtime/blob/9ee63fc65a9720568f74df9901b16883621636b4/pkg/cache/internal/informers_map.go#L223

What we do elsewhere in this situation is using `Eventually`, which polls: https://github.com/kubernetes-sigs/controller-runtime/blob/9ee63fc65a9720568f74df9901b16883621636b4/pkg/cache/cache_test.go#L1142

That is also not ideal either ofc but IMHO better than a sleep, because depending on the performance of the computer that executes these tests, the required duration of the `sleep` varies which effectively means that you have to use the worst case duration which in turn artificially slows down the tests on faster computers.

Reply:
I've replaced the sleeps with eventually/consistently usages. Marking as resolved.